### DISCUSSION TITLE
Getting the actual XPATH of an xml_node (xml_node::path() doesn't seem to do it...)

### DISCUSSION THREAD

**Discussion by tstack** [zeux/pugixml#396] [Q&A] [answered]
Hi,

I'm working on a SQLite extension that provides an `xpath()` table-valued function and wanted to return the "actual" path of a node in the `xpath_node_set`.  I was hoping to use `xml_node::path()`, but that doesn't seem to return what I was expecting.  For example, given this document:

```xml
<abc>
    <def>
        Hi
    </def>
    <def>
        Bye
    </def>
</abc>
```

The xpath expression `/abc/def` returns the expected nodes, however, the `path()` method returns `/abc/def` for both nodes.  I kind of expect the following paths to be returned: `/abc/def[1]` and `/abc/def[2]`.  Basically, I think you should be able to use the result of `path()` in an XPATH expression to get back to that very node.

So, is this the expected behavior of `path()` or would you consider it a bug?

Thanks,

Tim

**Comment by zeux** [comment, marked as answer]
`path()` predates XPath, and uses a simpler syntax that is compatible with `first_element_by_path` - but it indeed doesn't work when the sibling name is not unique. It's not a bug, just a limitation of the format that `path()` works with; maybe it's worth highlighting in the documentation / header comment, but otherwise it's by design.

Note that the general construction of the path may need to be somewhat domain-specific, for example if the path needs to survive document changes but elements have unique ids then it may be better to construct smth like `/abc/def[@id = 'hi']`.

**Reply by tstack** [reply]
Sounds good.  Would you be open to a PR that does return the "actual" XPATH for an xml_node?
